train/scripts/train_xpu_ed.py

```diff
@@ -0,0 +1,190 @@
+except ValueError:
+    # We must be running on a single node.
+    master_addr = "0.0.0.0"
+
+os.environ["MASTER_ADDR"] = master_addr
+os.environ["MASTER_PORT"] = "29876"
+USE_SUBDEVICES = os.environ.get("USE_SUBDEVICES", False)
+
+def main(download_path: str):
+    comms_backend = "ccl"
+    device_type = "xpu"
+
+    time_start_total = time.time()
+
+    print("Initialising process group with backend", comms_backend, flush=True)
+    # ToDo Run 2 or more processes.
+    init_process_group(
+        world_size=int(WORLD_SIZE),
+        rank=int(RANK),
+        backend=comms_backend,
+    )
+
+    device = f"{device_type}:{LOCAL_RANK}"
+    print(f"Using {device=}", flush=True)
+
+    print("loading model...", flush=True)
+    model = Aurora(
+        use_lora=False,  # Model was not fine-tuned.
+        autocast=True,  # Use AMP.
+        encoder_depths=(12,12,12),
+        encoder_num_heads=(4, 8, 16),
+        decoder_depths=(12,12,12),
+        decoder_num_heads=(16, 8, 4),
+        embed_dim=256,
+        num_heads=8,
+    )
+    # can no longer load checkpoint as we have different model size
+    # model.load_checkpoint("microsoft/aurora", "aurora-0.25-pretrained.ckpt")
+
+    download_path = Path(download_path)
+
+    print("preparing model...", flush=True)
+    model.configure_activation_checkpointing()
+    model = FSDP(
+        model,
+        device_id=LOCAL_RANK,
+        use_orig_params=True,
+        sharding_strategy=ShardingStrategy.NO_SHARD,
+    )
+    model.train()
+
+    # AdamW, as used in the paper.
+    optimizer = torch.optim.AdamW(model.parameters())
+
+    print("loading data...", flush=True)
+    dataset = AuroraDataset(
+        data_path=download_path,
+        t=1,
+        static_data=Path("static.nc"),
+        surface_data=Path("2023-01-surface-level.nc"),
+        atmos_data=Path("2023-01-atmospheric.nc"),
+    )
+    sampler = DistributedSampler(dataset)
+
+    data_loader = DataLoader(
+        dataset=dataset,
+        batch_size=1,  # If we set a batch size we'll need a collate_fn
+        shuffle=False,  # We don't need to shuffle.
+        sampler=sampler,
+        collate_fn=aurora_collate_fn,
+    )
+
+    times = []
+
+    time_start = time.time()
+    for batch, (X, y) in enumerate(data_loader):
+        print(f"batch {batch}...", flush=True)
+
+        optimizer.zero_grad()
+
+        with torch.autocast(device_type=device_type):
+            print("performing forward pass...", flush=True)
+            pred = model(X)
+
+            # only one of these is necessary
+            pred = pred.to(device)
+            y = y.to(device)
+
+            # mean absolute error of one variable
+            print("calculating loss...", flush=True)
+
+            # Todo: Are pred's of type PyTree and does it matter?
+            loss = mae(pred, y)
+
+        print("performing backward pass...", flush=True)
+        loss.backward()
+
+        print("optimizing...", flush=True)
+        optimizer.step()
```

Comment:
One quick question. In the final timings elsewhere I think you did this:

```suggestion
        if batch % n_batches_per_optim == 0:
            print("optimizing...")
            optimizer.step()
```

With:
```python
    n_batches_per_optim = 8 / WORLD_SIZE
```

Should this be the case here as well? Apologies for not raising this earlier, but I thought maybe this had been superseded by the `batch_size` change.

Reply:
We talked about this earlier but I'm adding here so its written down, we didn't do this for this bit of code but could do. I think the effort of changing it is maybe not worth it.